sklearn/metrics/_classification.py

```diff
@@ -2457,3 +2457,135 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
             pos_label = y_true.max()
     y_true = np.array(y_true == pos_label, int)
     return np.average((y_true - y_prob) ** 2, weights=sample_weight)
+
+
+def calibration_loss(y_true, y_prob, sample_weight=None, norm="l2",
+                     n_bins=10, pos_label=None, reduce_bias=True):
+    """Compute calibration loss.
+
+    Across all items in a set of N predictions, the calibration loss measures
+    the aggregated difference between (1) the average predicted probabilities
+    assigned to the positive class, and (2) the frequencies
+    of the positive class in the actual outcome.
+
+    The calibration loss is appropriate for binary and categorical outcomes
+    that can be structured as true or false.
+    Which label is considered to be the positive label is controlled via the
+    parameter pos_label, which defaults to 1.
+
+    Read more in the :ref:`User Guide <calibration>`.
+
+    Parameters
+    ----------
+    y_true : array, shape (n_samples,)
+        True targets.
+
+    y_prob : array, shape (n_samples,)
+        Probabilities of the positive class.
+
+    sample_weight : array-like, shape (n_samples,), optional
+        Sample weights.
+
+    norm : 'l1' | 'l2' | 'max'
+        Norm method.
+
+    n_bins : int, optional (default=10)
+       The number of bins to compute error on.
+
+    pos_label : int or str, optional (default=None)
+        Label of the positive class. If None, the maximum label is used as
+        positive class
+
+    reduce_bias : bool, optional (default=True)
+        Add debiasing term as in Verified Uncertainty Calibration, A. Kumar
+
+    Returns
+    -------
+    score : float
+        calibration loss
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics import calibration_loss
+    >>> y_true = np.array([0, 0, 0, 1] + [0, 1, 1, 1])
+    >>> y_pred = np.array([0.25, 0.25, 0.25, 0.25] + [0.75, 0.75, 0.75, 0.75])
+    >>> calibration_loss(y_true, y_pred, n_bins=5)
+    0.0
+    >>> calibration_loss(y_true, y_pred, n_bins=5, \
+                         norm="max")
+    0.0
+    >>> y_true = np.array([0, 0, 0, 0] + [1, 1, 1, 1])
+    >>> calibration_loss(y_true, y_pred, n_bins=5)
+    0.25
+    >>> calibration_loss(y_true, y_pred, n_bins=5, \
+                         norm="max")
+    0.25
+    """
+    y_true = column_or_1d(y_true)
+    y_prob = column_or_1d(y_prob)
+    assert_all_finite(y_true)
+    assert_all_finite(y_prob)
+    check_consistent_length(y_true, y_prob, sample_weight)
+    if any(y_prob < 0) or any(y_prob > 1):
+        raise ValueError("y_prob has values outside of [0, 1] range")
+
+    norm_options = ('l1', 'l2', 'max')
+    if norm not in norm_options:
+        raise ValueError('norm has to be one of ' +
+                         str(norm_options))
+
+    n_bins = int(n_bins)
+
+    if pos_label is None:
+        pos_label = y_true.max()
+    y_true = np.array(y_true == pos_label, int)
```

Comment:
This should be okay now with [what's above](https://github.com/scikit-learn/scikit-learn/blob/3c2d66241712246ff6c6cec088b24c3e2ab1d89a/sklearn/metrics/_classification.py#L2542-L2552) ?
But there still a test to be added to raise that `ValueError`

Reply:
Indeed. Although I am not sure that:

```python
    if pos_label is None:
        pos_label = y_true.max()
```

is a good idea. If `y_true.dtype.kind != i and pos_label is None` we should probably raise a value error to ask the user to pass `pos_label` explicitly no?

But then this makes things complicated to use `neg_calibration_error` as a scorer for `GridSearchCV` when the target variable `y` is encoded with string labels.

This is kind of related to recently investigated issues in ROC and PR curves: #15573.

/cc @glemaitre .

Reply:
This is related to https://github.com/scikit-learn/scikit-learn/pull/17594#issuecomment-648817284 for `roc_auc_score`.

Reply:
@dsleo maybe it is worth noting somewhere that this function doesn't support `y_true` to be string? I think if it a string, we would get a:
```
TypeError: cannot perform reduce with flexible type
``` 
error when calling `max()`. Maybe we could give a more informative error for the user here?

Reply:
But I am not sure what the resolution for @ogrisel's comment was. I assume it was decided that string `y_true` would not be supported?

Reply:
If the score is not symmetric, I would make `pos_label` consistent with the scorer that raise an error with string.
To be more specific, we should carry this check:

https://github.com/scikit-learn/scikit-learn/blob/e89157e18a47378a948b8f4783507f7e70b11f71/sklearn/metrics/_ranking.py#L553-L572

We could make a small refactoring indeed. So `pos_label` would be in line with:

```
pos_label: int or str, default=None

    The label of the positive class. When pos_label=None, if y_true is in {-1, 1} or {0, 1},
    pos_label is set to 1, otherwise, an error will be raised.

```